### PR TITLE
Fix calling codes in registration flow.

### PR DESCRIPTION
**Pull request by charlesmchen**
Fixes: https://github.com/WhisperSystems/Signal-iOS/issues/1821

PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/view controllers/CountryCodeViewController.m [hunk 1]
```diff
@@ -18,10 +18,69 @@ @implementation CountryCodeViewController
...
+            ![callingCode isEqualToString:@"+0"]) {
+            [result addObject:countryCode];
+        }
+    }
+    return result;
+}
+
+- (NSString *)callingCodeFromCountryCode:(NSString *)countryCode {
+    NSString *callingCode = [PhoneNumberUtil callingCodeFromCountryCode:countryCode];
+    if ([countryCode isEqualToString:@"AQ"]) {
```
Comment (charlesmchen) [inline Signal/src/view controllers/CountryCodeViewController.m]: `PhoneNumberUtil` uses a third-party library `NBPhoneNumberUtil`, which is returning bad data for some obscure country codes.

File: Signal/src/view controllers/CountryCodeViewController.m [hunk 2]
```diff
@@ -18,10 +18,69 @@ @implementation CountryCodeViewController
...
     self.title    = NSLocalizedString(@"COUNTRYCODE_SELECT_TITLE", @"");
 }
 
+- (NSArray *)countryCodesForSearchTerm:(NSString *)searchTerm {
+    NSMutableArray *result = [NSMutableArray array];
+    for (NSString *countryCode in [PhoneNumberUtil countryCodesForSearchTerm:nil]) {
+        NSString *callingCode = [self callingCodeFromCountryCode:countryCode];
+        if (callingCode != nil &&
+            ![callingCode isEqualToString:@"+0"]) {
+            [result addObject:countryCode];
```
Comment (charlesmchen) [inline Signal/src/view controllers/CountryCodeViewController.m]: We fix the countries that we can, but we also do this filtering for countries that we can't fix.  This future-proofs us as well.

File: Signal/src/view controllers/CountryCodeViewController.m [hunk 3]
```diff
@@ -18,10 +18,69 @@ @implementation CountryCodeViewController
...
+    } else if ([countryCode isEqualToString:@"IC"]) {
+        // Canary Islands
+        callingCode = @"+34";
+    } else if ([countryCode isEqualToString:@"EA"]) {
+        // Ceuta & Melilla
+        callingCode = @"+34";
+    } else if ([countryCode isEqualToString:@"CP"]) {
+        // Clipperton Island
+        //
+        // This country code should be filtered.
```
Comment (michaelkirk) [inline Signal/src/view controllers/CountryCodeViewController.m]: Can you add a comment as to why it should be filtered, and for Heard & MCDonald islands?

Comment (charlesmchen) [inline Signal/src/view controllers/CountryCodeViewController.m]: Done. Essentially, I can't find reliable info on their calling codes.  I've added the most probable value for the Heard & McDonald Islands, but I'm going to filter Clipperton Island which appears to be uninhabited and not to have a valid calling code.

**Comment by michaelkirk** [comment]
rebased and merged with https://github.com/WhisperSystems/Signal-iOS/commit/4983853dd05b6bbbef8b1325aebb9fa0ce477343

**Pull request closed without merging**